### PR TITLE
feat: S2I strategy for on cluster build

### LINKED ISSUE #989: On-cluster build: s2i support

### PR DESCRIPTION
**Pull request by zroubalik**
Signed-off-by: Zbynek Roubalik <zroubalik@gmail.com>

Using `func-s2i` Tekton Task as a temporary solution, till my changes in the upstream will be accepted: https://github.com/tektoncd/catalog/pull/1055

# Changes

- :gift: S2I strategy for on cluster build

The S2I Tekton Task that's been used by this feature needs changes, PR opened: https://github.com/tektoncd/catalog/pull/1025

/kind enhancement

Fixes #989

### CODE REVIEW

File: s2i/builder.go [hunk 1]
```diff
@@ -103,7 +103,7 @@ func NewBuilder(options ...Option) *Builder {
 func (b *Builder) Build(ctx context.Context, f fn.Function) (err error) {
 	// TODO this function currently doesn't support private s2i builder images since credentials are not set
 
-	// Builder image from the function if defined, default otherwise.
+	// Builder image from the function  if defined, default otherwise.
```
Comment (lance) [inline s2i/builder.go:106]: You can probably stash this change

**Review by lance — COMMENTED** [review]
Overall lgtm. A couple of questions. I see you now have tests for additional buildpacks for the pack builds. But go and rust are still unsupported. Has the groundwork been laid to handle our external buildpacks in rust and go? Also, we are now handing env vars (thanks!) but I suppose labels and annotations from func.yaml are not being applied?

**Comment by zroubalik** [comment]
> Overall lgtm. A couple of questions. I see you now have tests for additional buildpacks for the pack builds. But go and rust are still unsupported. Has the groundwork been laid to handle our external buildpacks in rust and go? Also, we are now handing env vars (thanks!) but I suppose labels and annotations from func.yaml are not being applied?

- Go and Rust are still unsuported, no relevant change in this PR, because additional Buildpacks are not supported yet :(
- Labels from `func.yaml` are applied here to all Pipeline resources: https://github.com/knative-sandbox/kn-plugin-func/blob/e6ec11b0e5f59062dc25a234f146fabef74cb7e5/pipelines/tekton/pipeplines_provider.go#L100
- Annotations are not applied, but that could be done: https://github.com/knative-sandbox/kn-plugin-func/issues/1192

**Comment by lance** [comment]
/lgtm
Thanks @zroubalik!

**Pull request merged**